### PR TITLE
gh-1016 Added functionality for the settings

### PR DESCRIPTION
**Pull request by DmitryBespalov**
- Biometry enable/disable
- App login on/off
- Confirmation on/off

### CODE REVIEW

File: Multisig/Logic/AuthenticationController.swift [hunk 1]
```diff
@@ -102,6 +109,29 @@ class AuthenticationController {
         }
         return Data(derivedKey).toHexString()
     }
+
+
+    // MARK: - Biometry
+
+    /// Is device hardware supports the biometry
+    var isBiometricsSupported: Bool {
```
Comment (sche) [inline Multisig/Logic/AuthenticationController.swift:117]: I would be for adding it to the lib, but okay :)

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift [hunk 2]
```diff
@@ -80,42 +86,105 @@ class PasscodeSettingsViewController: UITableViewController {
...
         }
-
-        present(nav, animated: true, completion: nil)
     }
 
-    private func changePasscode() {
+    private func withPasscodeAuthentication(
+        for reason: String,
+        tracking: TrackingEvent? = nil,
+        authenticated: @escaping (Bool, UINavigationController?, @escaping () -> Void) -> Void
```
Comment (sche) [inline Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift]: omg. This is a bit confusing

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift [hunk 3]
```diff
@@ -80,42 +86,105 @@ class PasscodeSettingsViewController: UITableViewController {
...
+                }
+                nav?.pushViewController(changeVC, animated: true)
+            } else {
+                finish()
+            }
+        }
+    }
+
+    private func toggleUsage(option: PasscodeOptions) {
+        withPasscodeAuthentication(for: "Require to open app") { [unowned self] success, _, finish in
```
Comment (sche) [inline Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift]: should this string depend on the option?

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift [hunk 4]
```diff
@@ -80,42 +86,105 @@ class PasscodeSettingsViewController: UITableViewController {
...
+            }
+
+            finish()
+        }
+    }
+
+    private func toggleBiometrics() {
+        withPasscodeAuthentication(for: "Login with biometrics") { success, nav, finish in
+            if AppSettings.passcodeOptions.contains(.useBiometry) {
+                AppSettings.passcodeOptions.remove(.useBiometry)
```
Comment (sche) [inline Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift:148]: I remember we discussed it some time ago, but I have not found it in Passcode v2 requirements is that if action does not reduce security, it should not require confirmation with passcode. Disabling biometry actually increases security, so I would suggest not to ask confirmation for this case. I will add it to the refinement meeting agenda to discuss

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift (hunk 3 above)
Comment (DmitryBespalov) [reply Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift]: yes, good catch

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift (hunk 4 above)
Comment (DmitryBespalov) [reply Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift:148]: weren't you saying before that we should request biometry in any case when entering passcode?

Or you're saying that we should not request passcode for disabling biometry?

**Comment by sche** [comment]
Bug:
1. Try to enable biometry.
2. Close popover window
=> System suggestion to enable biometry appears

File: Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift (hunk 4 above)
Comment (sche) [reply Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift:148]: 2nd

**Comment by sche** [comment]
Bug: Disabling biometry in Phone settings for the app does not influence the switch
Steps:
1. Enable biometry
2. Go to phone settings and select the app
3. Disable biometry in global settings for the app
4. Navigate to app's passcode settings
Expected result: Login with biometrics option should be disabled

File: Multisig/Logic/AuthenticationController.swift (hunk 1 above)
Comment (DmitryBespalov) [reply Multisig/Logic/AuthenticationController.swift:117]: will move it after this is merged

File: Multisig/UI/Settings/Passcode/PasscodeViewController.swift [hunk 5]
```diff
@@ -57,6 +59,9 @@ class PasscodeViewController: UIViewController, UITextFieldDelegate {
         trackEvent(.userPasscodeSkipped)
     }
 
+    @IBAction func didTapBiometry(_ sender: Any) {
```
Comment (sche) [inline Multisig/UI/Settings/Passcode/PasscodeViewController.swift:62]: I would add a comment here that it is meant for override

**Review by sche — APPROVED** [review]

**Pull request merged**